LimitRange updates for Resource Requirements Requests

derekwaynecarr opened the pull request:
@bgrant0607 - take a look when you get a chance.

I will add support for LimitRequestRatio enforcement in a follow-on PR.

derekwaynecarr committed:
Document need to run generated deep copy

derekwaynecarr commented:
Think I have a logic error in the admission plug-in where I am only checking that Request < Max, and not also Limit < Max (if specified).  Will poke the PR again when I make that update.

bgrant0607 commented on docs/design/admission_control_limit_range.md:
```diff
@@ -73,8 +73,8 @@ type LimitRangeItem struct {
-  // DefaultRequests resource requests on this kind by resource name
-  DefaultRequests ResourceList `json:"defaultRequests,omitempty" description:"default resource requests values on this kind by resource name if omitted"`
+  // DefaultRequest resource requests on this kind by resource name
+  DefaultRequest ResourceList `json:"defaultRequest,omitempty" description:"default resource requests values on this kind by resource name if omitted"`
```
Why change the name? defaultRequests is somewhat more consistent.

derekwaynecarr replied:
At the time, it bothered me that we were not consistent with the rest of the object.

For example, I thought if we have DefaultRequests, we should also have Defaults.

Since we have Default now, I thought DefaultRequest made sense.

bgrant0607 replied:
Technically, Defaults should be DefaultLimits, or Defaults should be of type ResourceRequirements.

DefaultRequest is fine.

derekwaynecarr replied:
I agree on DefaultLimits, but did not want to make that change until v2
API.

On Wednesday, August 12, 2015, Brian Grant <notifications@github.com> wrote:

> In docs/design/admission_control_limit_range.md
> <https://github.com/kubernetes/kubernetes/pull/12492#discussion_r36936897>
> :
>
> > @@ -73,8 +73,8 @@ type LimitRangeItem struct {
> >    Min ResourceList `json:"min,omitempty" description:"min usage constraints on this kind by resource name"`
> >    // Default resource limits on this kind by resource name
> >    Default ResourceList `json:"default,omitempty" description:"default resource limits values on this kind by resource name if omitted"`
> > -  // DefaultRequests resource requests on this kind by resource name
> > -  DefaultRequests ResourceList `json:"defaultRequests,omitempty" description:"default resource requests values on this kind by resource name if omitted"`
> > +  // DefaultRequest resource requests on this kind by resource name
> > +  DefaultRequest ResourceList `json:"defaultRequest,omitempty" description:"default resource requests values on this kind by resource name if omitted"`
>
> Technically, Defaults should be DefaultLimits, or Defaults should be of
> type ResourceRequirements.
>
> DefaultRequest is fine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12492/files#r36936897>.
>

derekwaynecarr replied:
Is there a doc to track v2 changes yet?  I think Defaults as type
ResourceRequirements makes the most sense at that time.

On Wednesday, August 12, 2015, Derek Carr <decarr@redhat.com> wrote:

> I agree on DefaultLimits, but did not want to make that change until v2
> API.
>
> On Wednesday, August 12, 2015, Brian Grant <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> In docs/design/admission_control_limit_range.md
>> <https://github.com/kubernetes/kubernetes/pull/12492#discussion_r36936897>
>> :
>>
>> > @@ -73,8 +73,8 @@ type LimitRangeItem struct {
>> >    Min ResourceList `json:"min,omitempty" description:"min usage constraints on this kind by resource name"`
>> >    // Default resource limits on this kind by resource name
>> >    Default ResourceList `json:"default,omitempty" description:"default resource limits values on this kind by resource name if omitted"`
>> > -  // DefaultRequests resource requests on this kind by resource name
>> > -  DefaultRequests ResourceList `json:"defaultRequests,omitempty" description:"default resource requests values on this kind by resource name if omitted"`
>> > +  // DefaultRequest resource requests on this kind by resource name
>> > +  DefaultRequest ResourceList `json:"defaultRequest,omitempty" description:"default resource requests values on this kind by resource name if omitted"`
>>
>> Technically, Defaults should be DefaultLimits, or Defaults should be of
>> type ResourceRequirements.
>>
>> DefaultRequest is fine.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/12492/files#r36936897>.
>>
>

bgrant0607 replied:
Agreed that it must be v2. #8190 is the umbrella issue.

I'm not going to fork any v2 APIs until API groups are ready.

davidopp replied:
s/if omitted/if resource request is omitted/

For LimitRequestRatio comment on next line it is clearer if you say "LimitRequestRatio is the maximum allows ratio of limit over request, representing the maximum burst for the named resource" or something like that. As-is, it's confusing.

bgrant0607 commented:
@davidopp PTAL.

derekwaynecarr commented:
This should be good to review now.

It took me a fair bit longer than I expected to test this, but I think the unit tests are much stronger now.

derekwaynecarr commented:
@davidopp a friendly poke.  Would like to get this in before I head out for vacation next week.

davidopp commented on api/swagger-spec/v1.json:
```diff
@@ -11540,6 +11540,14 @@
+     },
+     "defaultRequest": {
+      "type": "any",
+      "description": "default resource requests values on this kind by resource name if omitted"
```
I saw your discussion with @bgrant0607 later about not changing Default to DefaultLimit, but can you at least update the description of "default" above to "default resource limit values on this kind..." ? Or is that considered an API change?

davidopp commented on api/swagger-spec/v1.json:
```diff
@@ -11540,6 +11540,14 @@
+     },
+     "limitRequestRatio": {
+      "type": "any",
+      "description": "the ratio of limit over request that is the maximum allowed burst for the named resource. if specified, the named resource must have a request and limit that are both non-zero where limit divided by request is less than or equal to the enumerated value"
```
I would switch the order of these two sentences, as the second one is more important than the first one. In fact once you switch them, I would rewrite the new second one to something like "This represents the max burst for the named resource."

Also, I would suggest renaming to maxLimitRequestRatio to make this more self-describing.

davidopp commented on docs/design/admission_control_limit_range.md:
```diff
@@ -73,8 +73,8 @@ type LimitRangeItem struct {
   Default ResourceList `json:"default,omitempty" description:"default resource limits values on this kind by resource name if omitted"`
-  // DefaultRequests resource requests on this kind by resource name
-  DefaultRequests ResourceList `json:"defaultRequests,omitempty" description:"default resource requests values on this kind by resource name if omitted"`
+  // DefaultRequest resource requests on this kind by resource name
```
This comment is confusing (I see you tried to make it parallel to the one for Default but it doesn't really work). Better would be just "Default resource requests on this kind..." If there is some style rule that you have to start it with the name of the field, then I would say "DefaultRequest is default resource requests on this kind..."

Also in both this line and the Default line I would say "if [resource limits/resource requests] is omitted" to make it clear what you're talking about.

davidopp commented on pkg/api/testing/fuzzer.go:
```diff
@@ -174,6 +174,25 @@ func FuzzerFor(t *testing.T, version string, src rand.Source) *fuzz.Fuzzer {
 			q.Limits[api.ResourceStorage] = *storageLimit.Copy()
 			q.Requests[api.ResourceStorage] = *storageLimit.Copy()
 		},
+		func(q *api.LimitRangeItem, c fuzz.Continue) {
```
include LimitRequestRatio?

davidopp commented on pkg/api/types.go:
```diff
@@ -1940,6 +1940,10 @@ type LimitRangeItem struct {
 	Min ResourceList `json:"min,omitempty"`
 	// Default usage constraints on this kind by resource name
 	Default ResourceList `json:"default,omitempty"`
+	// DefaultRequest resource requests on this kind by resource name
```
See my earlier comment about making this comment clearer.

davidopp commented on pkg/api/types.go:
```diff
@@ -1940,6 +1940,10 @@ type LimitRangeItem struct {
 	Default ResourceList `json:"default,omitempty"`
+	// DefaultRequest resource requests on this kind by resource name
+	DefaultRequest ResourceList `json:"defaultRequest,omitempty"`
+	// LimitRequestRatio is the ratio of limit over request that is the maximum allowed burst for the named resource
```
See my earlier comment about this comment.

davidopp commented on pkg/api/v1/types.go:
```diff
@@ -1842,6 +1842,10 @@ type LimitRangeItem struct {
 	Min ResourceList `json:"min,omitempty" description:"min usage constraints on this kind by resource name"`
 	// Default usage constraints on this kind by resource name
 	Default ResourceList `json:"default,omitempty" description:"default values on this kind by resource name if omitted"`
+	// DefaultRequests resource requests on this kind by resource name
```
see earlier comment about this comment

davidopp commented on pkg/api/v1/types.go:
```diff
@@ -1842,6 +1842,10 @@ type LimitRangeItem struct {
 	Default ResourceList `json:"default,omitempty" description:"default values on this kind by resource name if omitted"`
+	// DefaultRequests resource requests on this kind by resource name
+	DefaultRequest ResourceList `json:"defaultRequest,omitempty" description:"default resource requests values on this kind by resource name if omitted"`
+	// LimitRequestRatio is the ratio of limit over request that is the maximum allowed burst for the named resource
```
see earlier comment about this comment

davidopp commented on pkg/api/v1/types.go:
```diff
@@ -1842,6 +1842,10 @@ type LimitRangeItem struct {
+	// DefaultRequests resource requests on this kind by resource name
+	DefaultRequest ResourceList `json:"defaultRequest,omitempty" description:"default resource requests values on this kind by resource name if omitted"`
+	// LimitRequestRatio is the ratio of limit over request that is the maximum allowed burst for the named resource
+	LimitRequestRatio ResourceList `json:"limitRequestRatio,omitempty" description:"the ratio of limit over request that is the maximum allowed burst for the named resource. if specified, the named resource must have a request and limit that are both non-zero where limit divided by request is less than or equal to the enumerated value"`
```
I would flip these sentences.

davidopp commented on pkg/api/validation/validation.go:
```diff
@@ -1426,28 +1427,62 @@ func ValidateLimitRange(limitRange *api.LimitRange) errs.ValidationErrorList {
 			defaultValue, defaultValueFound := defaults[k]
+			defaultRequestValue, defaultRequestValueFound := defaultRequests[k]
 
+			// max > min
```
I think if you're referring to the condition that causes an error, the comment should say min > max.
If you're referring to the condition that is accepted, the comment should say max >= min.

davidopp commented on pkg/api/validation/validation.go:
```diff
@@ -1426,28 +1427,62 @@ func ValidateLimitRange(limitRange *api.LimitRange) errs.ValidationErrorList {
+				allErrs = append(allErrs, errs.NewFieldInvalid(fmt.Sprintf("spec.limits[%d].min[%s]", i, k), minValue, fmt.Sprintf("min value %s is greater than max value %s", minQuantity.String(), maxQuantity.String())))
+			}
+
+			// request > min
```
min > request or request >= min depending on what you intended

davidopp commented on pkg/api/validation/validation.go:
```diff
@@ -1426,28 +1427,62 @@ func ValidateLimitRange(limitRange *api.LimitRange) errs.ValidationErrorList {
+				allErrs = append(allErrs, errs.NewFieldInvalid(fmt.Sprintf("spec.limits[%d].defaultRequest[%s]", i, k), defaultRequestValue, fmt.Sprintf("min value %s is greater than default request value %s", minQuantity.String(), defaultRequestQuantity.String())))
+			}
+
+			// request < min
```
request > max or request <= max depending on what you meant

davidopp commented on pkg/api/validation/validation.go:
```diff
@@ -1426,28 +1427,62 @@ func ValidateLimitRange(limitRange *api.LimitRange) errs.ValidationErrorList {
+				allErrs = append(allErrs, errs.NewFieldInvalid(fmt.Sprintf("spec.limits[%d].defaultRequest[%s]", i, k), defaultRequestValue, fmt.Sprintf("default request value %s is greater than max value %s", defaultRequestQuantity.String(), maxQuantity.String())))
+			}
+
+			// request < default
```
request > default or request <= default

davidopp commented on pkg/api/validation/validation.go:
```diff
@@ -1426,28 +1427,62 @@ func ValidateLimitRange(limitRange *api.LimitRange) errs.ValidationErrorList {
+				allErrs = append(allErrs, errs.NewFieldInvalid(fmt.Sprintf("spec.limits[%d].defaultRequest[%s]", i, k), defaultRequestValue, fmt.Sprintf("default request value %s is greater than default limit value %s", defaultRequestQuantity.String(), defaultQuantity.String())))
 			}
 
+			// min < default
```
min > default or min <= default

davidopp commented on pkg/api/validation/validation.go:
```diff
@@ -1426,28 +1427,62 @@ func ValidateLimitRange(limitRange *api.LimitRange) errs.ValidationErrorList {
+				allErrs = append(allErrs, errs.NewFieldInvalid(fmt.Sprintf("spec.limits[%d].default[%s]", i, k), minValue, fmt.Sprintf("min value %s is greater than default value %s", minQuantity.String(), defaultQuantity.String())))
 			}
 
+			// max > default
```
default > max or max >= default

davidopp commented on plugin/pkg/admission/limitranger/admission.go:
```diff
@@ -181,91 +185,176 @@ func mergePodResourceRequirements(pod *api.Pod, defaultRequirements *api.Resourc
+	observedReqValue, observedLimValue, enforcedValue := requestLimitEnforcedValues(req, lim, enforced)
+
+	if !reqExists || (observedReqValue == int64(0)) {
+		return fmt.Errorf("%s max limit to request ratio per %s is %s, but no request is specified.", resourceName, limitType, enforced.String())
```
add "or request is 0" ?

davidopp commented on plugin/pkg/admission/limitranger/admission.go:
```diff
@@ -181,91 +185,176 @@ func mergePodResourceRequirements(pod *api.Pod, defaultRequirements *api.Resourc
+		return fmt.Errorf("%s max limit to request ratio per %s is %s, but no request is specified.", resourceName, limitType, enforced.String())
+	}
+	if !limExists || (observedLimValue == int64(0)) {
+		return fmt.Errorf("%s max limit to request ratio per %s is %s, but no limit is specified.", resourceName, limitType, enforced.String())
```
add "or limit is 0" ?

derekwaynecarr commented:
Thanks for the review @davidopp , I ran out of time responding, so will poke this PR again when I return from vacation in a week.  I agree with all of your feedback.

derekwaynecarr commented:
@davidopp - rebased, and made the requested review changes now that I am back from vacation.

1. Modified LimitRequestRatio to be MaxLimitRequestRatio per your comment.  @bgrant0607 - assume you are in agreement with @davidopp on that change request.
2. Updated the swagger descriptions for improved clarity.

Hope it all looks good to you now.

derekwaynecarr committed:
Use singular, make LimitRequestRatio MaxLimitRequestRatio

derekwaynecarr committed:
Add fields to LimitRange object

derekwaynecarr committed:
Update kubectl describer for LimitRange

derekwaynecarr committed:
Update admission control logic for LimitRange

derekwaynecarr committed:
Update salt deployment for new limit-range

derekwaynecarr commented:
Kicking shippable as it appeared to flake on go 1.3.

derekwaynecarr commented:
@k8s-bot test this please

derekwaynecarr commented:
@davidopp - friendly poke on this PR so i can proceed on the ResourceQuota updates after.

davidopp commented:
Will look at it shortly.

bgrant0607 commented:
Sorry, still flogging my email into submission. The API change LGTM, thanks.

bgrant0607 commented:
Feel free to poke me on IRC, btw, though I won't be available much of tomorrow.

davidopp commented:
I am looking at this now.

davidopp commented:
LGTM
